src/test/java/giis/eshopcontainers/e2e/functional/CatalogTests.java

```diff
@@ -0,0 +1,59 @@
+package giis.eshopcontainers.e2e.functional;
+
+import giis.eshopcontainers.e2e.functional.common.BaseLoggedClass;
+import giis.eshopcontainers.e2e.functional.common.ElementNotFoundException;
+import giis.eshopcontainers.e2e.functional.utils.Click;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+import static giis.eshopcontainers.e2e.functional.utils.Shopping.*;
+
+class CatalogTests extends BaseLoggedClass {
+    @Test
+    @DisplayName("AddProductsToBasket")
+    void addProductsToBasket() throws ElementNotFoundException {
+        log.debug("Checking that the product buttons are disabled before the user logs-in");
+        WebElement productCupButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[1]/form/input[1]"));
+        Assertions.assertEquals("esh-catalog-button is-disabled", productCupButton.getAttribute("class"), "Before Login the eShop product button was expected for being disabled and was enabled");
+        this.login();
+        WebElement loggedUser = driver.findElement(By.xpath("//*[@id=\"logoutForm\"]/section[1]/div"));
+        Assertions.assertEquals("alice", loggedUser.getText(), "The user logged is not the expected, the expected was alice and the obtained is " + loggedUser.getText());
+        log.debug("Adding the first element: NetCore Cup ");
+        productCupButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[1]/form/input[1]"));
+        Assertions.assertEquals("esh-catalog-button ", productCupButton.getAttribute("class"), "The eShop product button was expected for being enabled and was disabled");
+        Click.element(driver, waiter, productCupButton);
+        Assertions.assertEquals(1, getNumShoppingItems(driver, waiter), "The number of elements in the basket doesn't match");
+        log.debug("Adding Two elements more: Hoodie and CupPin");
+        WebElement productHoodieBlackButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[3]/form/input[1]"));
+        Click.element(driver, waiter, productHoodieBlackButton);
+        Assertions.assertEquals(2, getNumShoppingItems(driver, waiter));
+        WebElement productCupPinButton = driver.findElement(By.xpath("/html/body/div/div[3]/div[6]/form/input[1]"));
+        Click.element(driver, waiter, productCupPinButton);
+        Assertions.assertEquals(3, getNumShoppingItems(driver, waiter));
+        this.logout();
```

Comment:
I have a concern about the login/logout pattern. During the first test case review, I got the idea that the base class (BaseLoggedClass) was created for every test that requires a logged user. However, this case seems to be intended to browse to the basket page without a login, then loging, and then add to basket and logout. But what would happen if other test case does a similar scenario and fails in the middle? Could let the user logged in and then enter in this test with an already logged user?

Should the logout be inside a teardown method (in all tests) to ensure that this situation does not happens?

Reply:
Javier, currently the logout its also in the teardown method (if the test fails for any reason, it goes to the tear-down method that performs:
```
    @AfterEach
    void tearDown(TestInfo testInfo) throws ElementNotFoundException {
        log.info("Disposing user and releasing/closing browser for the test {}", testInfo.getDisplayName());
        if (isLogged) {
            log.debug("Logging out user: {}", this.userName);
            this.logout();
        }
    }
```